crates/ra_syntax/src/ast/edit.rs

```diff
@@ -334,6 +337,85 @@ impl ast::UseTree {
     }
 }
 
+impl ast::MatchArmList {
+    #[must_use]
+    pub fn append_arms(&self, items: impl Iterator<Item = ast::MatchArm>) -> ast::MatchArmList {
+        let mut res = self.clone();
+        res = res.strip_if_only_whitespace();
+        if !res.syntax().text().contains_char('\n') {
+            res = make_multiline(res);
+        }
+        items.for_each(|it| res = res.append_arm(it));
+        res
+    }
+
+    fn strip_if_only_whitespace(&self) -> ast::MatchArmList {
+        let mut iter = self.syntax().children_with_tokens().skip_while(|it| it.kind() != T!['{']);
+        iter.next(); // Eat the curly
+        let mut inner = iter.take_while(|it| it.kind() != T!['}']);
+        if !inner.clone().all(|it| it.kind() == WHITESPACE) {
+            return self.clone();
+        }
+        let start = match inner.next() {
+            Some(s) => s,
+            None => return self.clone(),
+        };
+        let end = match inner.last() {
+            Some(s) => s,
+            None => start.clone(),
+        };
+        let res = self.replace_children(start..=end, &mut iter::empty());
+        res
+    }
+
+    #[must_use]
+    pub fn remove_placeholder(&self) -> ast::MatchArmList {
+        let placeholder = self.arms().find(|arm| {
+            if let Some(ast::Pat::PlaceholderPat(_)) = arm.pat() {
+                return true;
+            }
+            false
+        });
+        if let Some(placeholder) = placeholder {
+            let s: SyntaxElement = placeholder.syntax().clone().into();
+            let e = s.clone();
+            self.replace_children(s..=e, &mut iter::empty())
+        } else {
+            self.clone()
+        }
+    }
+
+    #[must_use]
+    pub fn append_arm(&self, item: ast::MatchArm) -> ast::MatchArmList {
+        let r_curly = match self.syntax().children_with_tokens().find(|it| it.kind() == T!['}']) {
+            Some(t) => t,
+            None => return self.clone(),
+        };
+        let mut sib = r_curly.prev_sibling_or_token();
+        while let Some(s) = sib.clone() {
+            if let Some(tok) = s.as_token() {
+                if tok.kind() != WHITESPACE {
```

Comment:
What happened if it is a COMMENT ? maybe just use `kind().is_trival`  ?

Reply:
At the moment it just appends to the contents of the match, so if it's a comment the additional arms appear below it. I suppose that's a matter of taste as to whether to group the arms together or to have the arms truly appended. Currently:
```rust
enum A { One, Two}
match a {
     // foo
     A::One => {},
     // bar
}
```
gets changed to 
```rust
match a {
     // foo
     A::One => {}
     // bar
     A::Two => {}
}
```
using is_trivia would give:
```rust
match a {
      // foo
     A::One => {}
     A::Two => {}
     // bar
}
```

Happy to change if you prefer the latter :)

Reply:
From my viewpoint, we should literally append new code to the existing one instead of inserting in-between thus more likely distorting the original meaning of trivia already located in the match arms list. E.g. that `//bar` comment most likely could be focused on explaining the `A::One` arm then the closing curly brace (but even this is heuristic).

Reply:
Agreed ! But OTOH, why would we need to search fort the previous WHITESPACE ? Could we just InsertPosition::Before the `T!['}'] token instead ?

Reply:
Hmm, that is also heuristic. If we do that, the get the following behavior:
```rust
match a {
    A::One => {}

  
}
// transformed to:
match a {
    A::One => {}

 
    A::Two => {}
}
```
But if we skip whitespace we get the following:
```rust
match a {
    A::One => {}

  
}
// transformed to:
match a {
    A::One => {}
    A::Two => {}

 
    
}
```
Maybe we should not preserve literally all trivia, but maybe trim the trailing whitespace (so that we dont `strip_if_only_whitespace()`, but `strip_trailing_whitspace()`)?
Anyway, what we lack here is the `ra_fmt` crate that handles all this ad hoc formatting for us...
But, anyway, what do you think of my proposal?

Reply:
> we lack here is the ra_fmt crate that handles all this ad hoc formatting for us...
Yes, lack of `ra_fmt` crate is the main issue here.

Generally, conservative choices are better, so I would prefer to keep all the whitespaces as it is.

Reply:
> Agreed ! But OTOH, why would we need to search for previous WHITESPACE ? Could we just `InsertPosition::Before` the `T!['}']` token instead ?

I think this was the easiest way of keeping the closing curly brace in the same position as we found it. I think without it the logic for inserting the extra arms was more complicated, but I'll have to refresh my memory to be sure!